File: crates/server/src/protocol.rs
use tokio::io::{AsyncReadExt, AsyncWriteExt};

const MSG_CONNECT: u8     = 0x01;
const MSG_CONNECT_OK: u8  = 0x02;
const MSG_QUERY: u8       = 0x03;
const MSG_RESULT_ROWS: u8 = 0x07;
const MSG_RESULT_SCALAR: u8 = 0x08;
const MSG_RESULT_OK: u8   = 0x09;
const MSG_ERROR: u8       = 0x0A;
const MSG_DISCONNECT: u8  = 0x10;

/// Maximum payload size accepted from the wire (64 MB).
const MAX_PAYLOAD_SIZE: usize = 64 * 1024 * 1024;

/// Maximum number of columns allowed in a result set.
const MAX_COLUMNS: usize = 4096;

/// Maximum number of rows allowed in a single result message.
const MAX_ROWS: usize = 10_000_000;

#[derive(Debug, Clone)]
pub enum Message {
    Connect { db_name: String, password: Option<String> },
    ConnectOk { version: String },
    Query { query: String },
    ResultRows {
        columns: Vec<String>,
        rows: Vec<Vec<String>>,
    },
    ResultScalar { value: String },
    ResultOk { affected: u64 },
    Error { message: String },
    Disconnect,
}

impl Message {
    /// Encode message into wire format: [type(1)][flags(1)][len(4)][payload]
    pub fn encode(&self) -> Vec<u8> {
        let (msg_type, payload) = match self {
            Message::Connect { db_name, password } => {
                let mut buf = encode_string(db_name);
                // Password is encoded as a length-prefixed string. Empty (len=0) means None.
                match password {
                    Some(p) => buf.extend_from_slice(&encode_string(p)),
                    None => buf.extend_from_slice(&0u32.to_le_bytes()),
                }
                (MSG_CONNECT, buf)
            }
            Message::ConnectOk { version } => (MSG_CONNECT_OK, encode_string(version)),
            Message::Query { query } => (MSG_QUERY, encode_string(query)),
            Message::ResultRows { columns, rows } => {
                let mut buf = Vec::new();
                buf.extend_from_slice(&(columns.len() as u16).to_le_bytes());
                for col in columns {
                    buf.extend_from_slice(&encode_string(col));
                }
                buf.extend_from_slice(&(rows.len() as u32).to_le_bytes());
                for row in rows {
                    for val in row {
                        buf.extend_from_slice(&encode_string(val));
                    }
                }
                (MSG_RESULT_ROWS, buf)
            }
            Message::ResultScalar { value } => (MSG_RESULT_SCALAR, encode_string(value)),
            Message::ResultOk { affected } => (MSG_RESULT_OK, affected.to_le_bytes().to_vec()),
            Message::Error { message } => (MSG_ERROR, encode_string(message)),
            Message::Disconnect => (MSG_DISCONNECT, Vec::new()),
        };

        let mut frame = Vec::with_capacity(6 + payload.len());
        frame.push(msg_type);
        frame.push(0); // flags
        frame.extend_from_slice(&(payload.len() as u32).to_le_bytes());
        frame.extend_from_slice(&payload);
        frame
    }

    /// Decode message from wire format.
    pub fn decode(data: &[u8]) -> Result<Message, String> {
        if data.len() < 6 {
            return Err("frame too short".into());
        }
        let msg_type = data[0];
        let _flags = data[1];
        let len_bytes: [u8; 4] = data[2..6]
            .try_into()
            .map_err(|_| "invalid header length field".to_string())?;
        let payload_len = u32::from_le_bytes(len_bytes) as usize;
        if 6 + payload_len > data.len() {
            return Err("payload length exceeds frame".into());
        }
        let payload = &data[6..6 + payload_len];

        match msg_type {
            MSG_CONNECT => {
                let mut pos = 0;
                let db_name = decode_string(payload, &mut pos)?;
                // Password is optional. If there are no more bytes, treat as None
                // (backwards compatible with old clients that don't send a password).
                let password = if pos < payload.len() {
                    let p = decode_string(payload, &mut pos)?;
                    if p.is_empty() { None } else { Some(p) }
                } else {
                    None
                };
                Ok(Message::Connect { db_name, password })
            }
            MSG_CONNECT_OK => {
                let version = decode_string(payload, &mut 0)?;
                Ok(Message::ConnectOk { version })
            }
            MSG_QUERY => {
                let query = decode_string(payload, &mut 0)?;
                Ok(Message::Query { query })
            }
            MSG_RESULT_ROWS => {
                let mut pos = 0;
                if pos + 2 > payload.len() {
                    return Err("truncated column count".into());
                }
                let col_bytes: [u8; 2] = payload[pos..pos+2]
                    .try_into()
                    .map_err(|_| "invalid column count bytes".to_string())?;
                let col_count = u16::from_le_bytes(col_bytes) as usize;
                pos += 2;
                if col_count > MAX_COLUMNS {
                    return Err("too many columns".into());
                }
                let mut columns = Vec::with_capacity(col_count);
                for _ in 0..col_count {
                    columns.push(decode_string(payload, &mut pos)?);
                }
                if pos + 4 > payload.len() {
                    return Err("truncated row count".into());
                }
                let row_bytes: [u8; 4] = payload[pos..pos+4]
                    .try_into()
                    .map_err(|_| "invalid row count bytes".to_string())?;
                let row_count = u32::from_le_bytes(row_bytes) as usize;
                pos += 4;
                if row_count > MAX_ROWS {
                    return Err("too many rows".into());
                }
                let mut rows = Vec::with_capacity(row_count);
                for _ in 0..row_count {
                    let mut row = Vec::with_capacity(col_count);
                    for _ in 0..col_count {
                        row.push(decode_string(payload, &mut pos)?);
                    }
                    rows.push(row);
                }
                Ok(Message::ResultRows { columns, rows })
            }
            MSG_RESULT_SCALAR => {
                let value = decode_string(payload, &mut 0)?;
                Ok(Message::ResultScalar { value })
            }
            MSG_RESULT_OK => {
                if payload.len() < 8 {
                    return Err("truncated result ok payload".into());
                }
                let aff_bytes: [u8; 8] = payload[0..8]
                    .try_into()
                    .map_err(|_| "invalid affected count bytes".to_string())?;
                let affected = u64::from_le_bytes(aff_bytes);
                Ok(Message::ResultOk { affected })
            }
            MSG_ERROR => {
                let message = decode_string(payload, &mut 0)?;
                Ok(Message::Error { message })
            }
            MSG_DISCONNECT => Ok(Message::Disconnect),
            _ => Err(format!("unknown message type: {msg_type:#x}")),
        }
    }

    /// Write this message to an async writer.
    pub async fn write_to<W: AsyncWriteExt + Unpin>(&self, writer: &mut W) -> std::io::Result<()> {
        let bytes = self.encode();
        writer.write_all(&bytes).await
    }

    /// Read a message from an async reader.
    pub async fn read_from<R: AsyncReadExt + Unpin>(reader: &mut R) -> std::io::Result<Option<Message>> {
        let mut header = [0u8; 6];
        match reader.read_exact(&mut header).await {
            Ok(_) => {}
            Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => return Ok(None),
            Err(e) => return Err(e),
        }
        let len_bytes: [u8; 4] = header[2..6]
            .try_into()
            .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidData, "invalid header length field"))?;
        let payload_len = u32::from_le_bytes(len_bytes) as usize;
        if payload_len > MAX_PAYLOAD_SIZE {
            return Err(std::io::Error::new(
                std::io::ErrorKind::InvalidData,
                format!("payload too large: {payload_len} bytes (max {MAX_PAYLOAD_SIZE})"),
            ));
        }
        let mut payload = vec![0u8; payload_len];
        if payload_len > 0 {
            reader.read_exact(&mut payload).await?;
        }

        let mut full = Vec::with_capacity(6 + payload_len);
        full.extend_from_slice(&header);
        full.extend_from_slice(&payload);

        Message::decode(&full)
            .map(Some)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))
    }
}

fn encode_string(s: &str) -> Vec<u8> {
    let mut buf = Vec::with_capacity(4 + s.len());
    buf.extend_from_slice(&(s.len() as u32).to_le_bytes());
    buf.extend_from_slice(s.as_bytes());
    buf
}

fn decode_string(data: &[u8], pos: &mut usize) -> Result<String, String> {
    if *pos + 4 > data.len() {
        return Err("truncated string length".into());
    }
    let len_bytes: [u8; 4] = data[*pos..*pos+4]
        .try_into()
        .map_err(|_| "invalid string length bytes".to_string())?;
    let len = u32::from_le_bytes(len_bytes) as usize;
    *pos += 4;
    if *pos + len > data.len() {
        return Err("truncated string data".into());
    }
    let s = String::from_utf8_lossy(&data[*pos..*pos+len]).into_owned();
    *pos += len;
    Ok(s)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_encode_decode_query() {
        let msg = Message::Query {
            query: "User filter .age > 30".into(),
        };
        let bytes = msg.encode();
        let decoded = Message::decode(&bytes).unwrap();
        match decoded {
            Message::Query { query } => assert_eq!(query, "User filter .age > 30"),
            _ => panic!("expected Query"),
        }
    }

    #[test]
    fn test_encode_decode_result_rows() {
        let msg = Message::ResultRows {
            columns: vec!["name".into(), "age".into()],
            rows: vec![
                vec!["Alice".into(), "30".into()],
                vec!["Bob".into(), "25".into()],
            ],
        };
        let bytes = msg.encode();
        let decoded = Message::decode(&bytes).unwrap();
        match decoded {
            Message::ResultRows { columns, rows } => {
                assert_eq!(columns, vec!["name", "age"]);
                assert_eq!(rows.len(), 2);
            }
            _ => panic!("expected ResultRows"),
        }
    }

    #[test]
    fn test_encode_decode_error() {
        let msg = Message::Error { message: "table not found".into() };
        let bytes = msg.encode();
        let decoded = Message::decode(&bytes).unwrap();
        match decoded {
            Message::Error { message } => assert_eq!(message, "table not found"),
            _ => panic!("expected Error"),
        }
    }

    #[test]
    fn test_frame_length() {
        let msg = Message::Query { query: "User".into() };
        let bytes = msg.encode();
        assert!(bytes.len() >= 6);
        let payload_len = u32::from_le_bytes(bytes[2..6].try_into().unwrap()) as usize;
        assert_eq!(bytes.len(), 6 + payload_len);
    }
}

File: clients/java/src/main/java/dev/zvn/powdb/Protocol.java
package dev.zvn.powdb;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * PowDB wire protocol encoder/decoder.
 *
 * Frame format: {@code [type(1)][flags(1)][len(4 LE)][payload]}.
 * Strings are encoded as {@code [len(4 LE)][utf-8 bytes]}.
 */
public final class Protocol {

    public static final byte MSG_CONNECT = 0x01;
    public static final byte MSG_CONNECT_OK = 0x02;
    public static final byte MSG_QUERY = 0x03;
    public static final byte MSG_RESULT_ROWS = 0x07;
    public static final byte MSG_RESULT_SCALAR = 0x08;
    public static final byte MSG_RESULT_OK = 0x09;
    public static final byte MSG_ERROR = 0x0A;
    public static final byte MSG_DISCONNECT = 0x10;

    private Protocol() {}

    /** Carrier for decode results (the parsed message + bytes consumed). */
    public record Decoded(Message message, int consumed) {}

    public static byte[] encode(Message msg) {
        byte msgType;
        byte[] payload;
        switch (msg) {
            case Message.Connect c -> {
                msgType = MSG_CONNECT;
                byte[] dbBuf = encodeString(c.dbName());
                byte[] pwBuf = c.password() == null ? u32Le(0) : encodeString(c.password());
                payload = concat(dbBuf, pwBuf);
            }
            case Message.ConnectOk c -> {
                msgType = MSG_CONNECT_OK;
                payload = encodeString(c.version());
            }
            case Message.Query q -> {
                msgType = MSG_QUERY;
                payload = encodeString(q.query());
            }
            case Message.ResultRows r -> {
                msgType = MSG_RESULT_ROWS;
                ByteBuffer bb = ByteBuffer.allocate(estimateRowsSize(r)).order(ByteOrder.LITTLE_ENDIAN);
                bb.putShort((short) r.columns().size());
                for (String c : r.columns()) appendString(bb, c);
                bb.putInt(r.rows().size());
                for (List<String> row : r.rows()) {
                    for (String v : row) appendString(bb, v);
                }
                payload = new byte[bb.position()];
                bb.flip();
                bb.get(payload);
            }
            case Message.ResultScalar s -> {
                msgType = MSG_RESULT_SCALAR;
                payload = encodeString(s.value());
            }
            case Message.ResultOk o -> {
                msgType = MSG_RESULT_OK;
                payload = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN).putLong(o.affected()).array();
            }
            case Message.Error e -> {
                msgType = MSG_ERROR;
                payload = encodeString(e.message());
            }
            case Message.Disconnect d -> {
                msgType = MSG_DISCONNECT;
                payload = new byte[0];
            }
        }
        byte[] frame = new byte[6 + payload.length];
        frame[0] = msgType;
        frame[1] = 0;
        ByteBuffer.wrap(frame, 2, 4).order(ByteOrder.LITTLE_ENDIAN).putInt(payload.length);
        System.arraycopy(payload, 0, frame, 6, payload.length);
        return frame;
    }

    /**
     * Tries to decode a single frame from the first {@code length} bytes of {@code buf}.
     * Returns {@code null} if more bytes are needed.
     *
     * @throws ProtocolException if the frame is malformed
     */
    public static Decoded tryDecode(byte[] buf, int length) {
        if (length < 6) return null;
        byte msgType = buf[0];
        int payloadLen = ByteBuffer.wrap(buf, 2, 4).order(ByteOrder.LITTLE_ENDIAN).getInt();
        if (payloadLen < 0) throw new ProtocolException("negative payload length");
        if (length < 6 + payloadLen) return null;
        ByteBuffer payload = ByteBuffer.wrap(buf, 6, payloadLen).order(ByteOrder.LITTLE_ENDIAN);
        return new Decoded(decodePayload(msgType, payload), 6 + payloadLen);
    }

    private static Message decodePayload(byte msgType, ByteBuffer p) {
        switch (msgType) {
            case MSG_CONNECT: {
                String db = decodeString(p);
                String pw = null;
                if (p.remaining() > 0) {
                    String s = decodeString(p);
                    pw = s.isEmpty() ? null : s;
                }
                return new Message.Connect(db, pw);
            }
            case MSG_CONNECT_OK:
                return new Message.ConnectOk(decodeString(p));
            case MSG_QUERY:
                return new Message.Query(decodeString(p));
            case MSG_RESULT_ROWS: {
                if (p.remaining() < 2) throw new ProtocolException("truncated column count");
                int colCount = Short.toUnsignedInt(p.getShort());
                List<String> cols = new ArrayList<>(colCount);
                for (int i = 0; i < colCount; i++) cols.add(decodeString(p));
                if (p.remaining() < 4) throw new ProtocolException("truncated row count");
                int rowCount = p.getInt();
                if (rowCount < 0) throw new ProtocolException("negative row count");
                List<List<String>> rows = new ArrayList<>(rowCount);
                for (int r = 0; r < rowCount; r++) {
                    List<String> row = new ArrayList<>(colCount);
                    for (int c = 0; c < colCount; c++) row.add(decodeString(p));
                    rows.add(row);
                }
                return new Message.ResultRows(cols, rows);
            }
            case MSG_RESULT_SCALAR:
                return new Message.ResultScalar(decodeString(p));
            case MSG_RESULT_OK: {
                if (p.remaining() < 8) throw new ProtocolException("truncated result ok payload");
                return new Message.ResultOk(p.getLong());
            }
            case MSG_ERROR:
                return new Message.Error(decodeString(p));
            case MSG_DISCONNECT:
                return new Message.Disconnect();
            default:
                throw new ProtocolException("unknown message type: 0x" + Integer.toHexString(msgType & 0xFF));
        }
    }

    private static byte[] encodeString(String s) {
        byte[] data = s.getBytes(StandardCharsets.UTF_8);
        byte[] out = new byte[4 + data.length];
        ByteBuffer.wrap(out, 0, 4).order(ByteOrder.LITTLE_ENDIAN).putInt(data.length);
        System.arraycopy(data, 0, out, 4, data.length);
        return out;
    }

    private static void appendString(ByteBuffer bb, String s) {
        byte[] data = s.getBytes(StandardCharsets.UTF_8);
        bb.putInt(data.length);
        bb.put(data);
    }

    private static String decodeString(ByteBuffer p) {
        if (p.remaining() < 4) throw new ProtocolException("truncated string length");
        int len = p.getInt();
        if (len < 0 || p.remaining() < len) throw new ProtocolException("truncated string data");
        byte[] data = new byte[len];
        p.get(data);
        return new String(data, StandardCharsets.UTF_8);
    }

    private static int estimateRowsSize(Message.ResultRows r) {
        int size = 2 + 4;
        for (String c : r.columns()) size += 4 + c.getBytes(StandardCharsets.UTF_8).length;
        for (List<String> row : r.rows()) {
            for (String v : row) size += 4 + v.getBytes(StandardCharsets.UTF_8).length;
        }
        return size;
    }

    private static byte[] u32Le(int n) {
        return ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN).putInt(n).array();
    }

    private static byte[] concat(byte[] a, byte[] b) {
        byte[] out = new byte[a.length + b.length];
        System.arraycopy(a, 0, out, 0, a.length);
        System.arraycopy(b, 0, out, a.length, b.length);
        return out;
    }
}
